test/lib/rules/no-this-property-during-ssr.js

```diff
@@ -0,0 +1,590 @@
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    if(document !== undefined) {
+                      doSomething(this.template);
+                    }
+                  }
+                }
+            `,
+        },
+    ],
+    invalid: [
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    this.template.querySelector('span')?.foo();
+                  }
+                }
+            `,
+            errors: [
+                {
+                    message:
+                        'You should not use `this.template.querySelector` in methods that will execute during SSR. Use `lwc:ref` instead.',
+                },
+            ],
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    this.foo();
+                  }
+                  foo() {
+                    this.template.querySelectorAll('span')?.foo();
+                  }
+                }
+            `,
+            errors: [
+                {
+                    message:
+                        'You should not use `this.template.querySelectorAll` in methods that will execute during SSR. Use `lwc:ref` instead.',
+                },
+            ],
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    this.foo();
+                  }
+                  foo() {
+                    doSomethingWith(this.template);
+                  }
+                }
+            `,
+            errors: [
+                {
+                    message:
+                        'You should not access any DOM properties on `this` in methods that will execute during SSR.',
+                },
+            ],
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    doSomethingWith(this.template);
+                  }
+                }
+            `,
+            errors: [
+                {
+                    message:
+                        'You should not access any DOM properties on `this` in methods that will execute during SSR.',
```

Comment:
> You should not access any DOM properties on `this` in methods that will execute during SSR.

Is this statement really true? As far as I know, we do implement at least `this.classList`, `this.getAttribute`, and `this.setAttribute`. Does it mean that we should retire those APIs?

Reply:
Reading is fine. Writing is not fine and causes hydration errors. An improvement would be to restrict the developer only from using APIs that change the DOM. Would you be okay with this going into a follow-up PR, @pmdartus?

Reply:
Sure. I don't see any issue with doing this work in a follow-up PR as long as we do it prior to socializing those new rules.